データベース設計

hmeg2896 opened the pull request:
# what
ER図とREADMEを作成

# why
データベースを実装するため

hmeg2896 committed:
データベース設計

chobi555 commented:
レビュー依頼ありがとうございます。
お手数ですが、ER図の画像をgyazoで撮影し、添付していただきたく存じます。
プルリクエスト内にgyazoのURLを添付していただくようお願い致します。

hmeg2896 commented:
https://gyazo.com/12021a21955fa73d87bdb5d4f4b6a9ea
ご確認の程お願いいたします。

yutomentors commented on README.md:
```diff
@@ -1,24 +1,79 @@
+| Column               | Type       | Options               |
+| -------------------- | ---------- | ----------------------|
+| user_name            | string     | null: false           |
+| email                | string     | null: false           |
```
```ruby
| email | string | null: false |
```
メールアドレスは一意性制約を持たせる必要がございます。
Optionsにその設定を加筆しましょう。

#### 理由：
要件に「メールアドレスは一意であること」とあるためです。

yutomentors commented on README.md:
```diff
@@ -1,24 +1,79 @@
+### アソシエーション
+has_many :items
+has_many :comments
+has_many :purchase
```
```
has_many :purchase
```
こちら　は複数形に変更しましょう。

理由：
アソシエーションの記述は
has_manyの場合、所有するモデルは複数形にしないと正しく動作しないためです。

yutomentors commented on README.md:
```diff
@@ -1,24 +1,79 @@
-* Services (job queues, cache servers, search engines, etc.)
+| column                 | Type          | options               |
+| ---------------------- | ------------- | ----------------------|
+| image                  | binary        | null: false           |
```
```ruby
| image | binary | null: false |
```
こちらのカラムは削除しましょう。
#### 理由：
今回の実装では商品出品の画像投稿において「active_storage」を使用いたします。
active_storage導入時に自動でテーブルなどが生成されるため、そちらに画像データが保存されるようになります。

yutomentors commented on README.md:
```diff
@@ -1,24 +1,79 @@
+| image                  | binary        | null: false           |
+| product                | string        | null: false           |
+| product_description    | text          | null: false           |
+| category               | string        | null: false           |
```
```
| category               | string        | null: false           |
| condition              | string        | null: false           |
| postage                | integer       | null: false           |
| region                 | string        | null: false           |
| shipping_date          | integer       | null: false           |
```

商品の状態・配送料の負担・発送元の地域・発送までの日数・カテゴリーに関してのカラムは
active_hashを用いての実装となります。

active_hashを用いたカラムはinteger型にしましょう！
また、idの値（数値）を保存するため、カラム名の最後に_idを付けるようにしましょう。

yutomentors commented on README.md:
```diff
@@ -1,24 +1,79 @@
 
-* Deployment instructions
+### アソシエーション
+belongs_to :users
```
「1つの商品は1つの購入情報を持っている」という設計にしたいため、
`has_one`を使用して購入管理テーブルとのアソシエーションを組みましょう。
#### 理由：
一つの商品は一つの購入履歴を持つという関係であるためです。

yutomentors replied:
```ruby
belongs_to :users
```
こちらのアソシエーションですが、単数形を見直してみましょう！
#### 理由：
Rails側で、アソシエーションを記述の際に、単数形・複数形のどちらで書くか決められているためです。

yutomentors commented on README.md:
```diff
@@ -1,24 +1,79 @@
+
+
+
+## purchaseテーブル
```
```
## purchaseテーブル
 ```

こちらテーブル名は複数形にしましょう！

chobi555 replied:
下記URLのコメント内容を実装しましょう。
https://github.com/hmeg2896/furima-36743/pull/14/files#r770404550

yutomentors commented on README.md:
```diff
@@ -1,24 +1,79 @@
+
+| column                 | Type          | options               |
+| ---------------------- | ------------- | ----------------------|
+| card_number            | integer       | null: false           |
```
```ruby
| card_number | integer | null: false |
| exp_month              | integer       | null: false           |
| exp_year               | integer       | null: false           |
| security_code          | integer       | null: false           |
```
カード情報の記述は削除しておきましょう！
必須機能の範囲ではカード情報は購入の都度入力するものなので、DBへの保存は不要であります。

#### 理由：
アプリケーションのサーバーサイドにおいて、クレジットカード情報を保持しないようにすることが義務付けられているためです。

chobi555 replied:
下記URLのコメント内容を実装しましょう。
https://github.com/hmeg2896/furima-36743/pull/14/files#r770404941

yutomentors commented on README.md:
```diff
@@ -1,24 +1,79 @@
+| exp_month              | integer       | null: false           |
+| exp_year               | integer       | null: false           |
+| security_code          | integer       | null: false           |
+| postal_code            | integer       | null: false           |
```
```ruby
| postal_code | integer | null: false |
```
こちらのカラムの型ですが、`string`型に変更しましょう。
#### 理由：
郵便番号はハイフンを含めて保存する仕様であり、文字列として扱うためです。

chobi555 replied:
下記URLのコメント内容を実装しましょう。
https://github.com/hmeg2896/furima-36743/pull/14/files#r770405237

yutomentors commented on README.md:
```diff
@@ -1,24 +1,79 @@
+| exp_year               | integer       | null: false           |
+| security_code          | integer       | null: false           |
+| postal_code            | integer       | null: false           |
+| prefecture             | string        | null: false           |
```
```ruby
| prefecture | string | null: false |
```
都道府県カラムはactive_hashを用いての実装となります。
active_hashを用いたカラムは、商品テーブルを参考にinteger型にしましょう！

#### 理由：
都道府県カラムはactive_hashを使用しての実装となります。
カラム名が同じであればactive_hashのデータは使い回すことが可能ですので、
都道府県カラムと商品テーブルの発送元カラムを同じ名前にしておくことが望ましいです。

chobi555 replied:
下記URLのコメント内容を実装しましょう。
https://github.com/hmeg2896/furima-36743/pull/14/files#r770405447

yutomentors commented on README.md:
```diff
@@ -1,24 +1,79 @@
+| prefecture             | string        | null: false           |
+| city                   | string        | null: false           |
+| address                | string        | null: false           |
+| building_name          | string        | null: false           |
```
```ruby
| building_name | string | null: false |
```
建物名は任意入力のデータとなります。
`null: false`は削除しましょう。

yutomentors commented on README.md:
```diff
@@ -1,24 +1,79 @@
+| city                   | string        | null: false           |
+| address                | string        | null: false           |
+| building_name          | string        | null: false           |
+| phone_number           | integer       | null: false
```
```ruby
| phone_number | integer | null: false
```
電話番号はstring型で作成しましょう。
#### 理由：
integer型の場合、0から始まる電話番号が上手く保存できなくなってしまうためです。

yutomentors commented on README.md:
```diff
@@ -1,24 +1,79 @@
+| address                | string        | null: false           |
+| building_name          | string        | null: false           |
+| phone_number           | integer       | null: false
+| user                   | references    | foreign_key: true     |  
```
```ruby
| user | references | foreign_key: true |
```
住所テーブルに、**購入履歴テーブル**の外部キーを保存するカラムに修正しましょう。

#### 理由：
必須機能の範囲では、住所情報を購入履歴のテーブルに紐付ける仕様であるためです。

chobi555 replied:
下記URLのコメント内容を実装しましょう。
https://github.com/hmeg2896/furima-36743/pull/14/files#r770406332

mihot117 replied:
```ruby
| user | references | foreign_key: true |
| order | references | foreign_key: true |
```

上記の記述で、購入履歴テーブルの外部キー付与を確認いたしました。
また、userカラムに関しては削除いただいて問題ございません。
75行目は消しておきましょう。

誰が・何を購入したのかは`ordersテーブル`で管理しています。
prdersテーブルの外部キーを取得した時点で
user, itemの情報は一括で紐づけられているため、user単独の外部キーを取得する必要はございません！

yutomentors commented on README.md:
```diff
@@ -1,24 +1,79 @@
+| user                   | references    | foreign_key: true     |  
+
+### アソシエーション
+belongs_to :users
```
```ruby
belongs_to :users
```
今回「１つの住所は購入情報のテーブルに紐づく」という設計にしたいため、
belongs_toを使用して購入管理テーブルとのアソシエーションを組みましょう。

yutomentors commented on README.md:
```diff
@@ -1,24 +1,79 @@
 
-This README would normally document whatever steps are necessary to get the
-application up and running.
 
```
購入履歴を保存するテーブルを作成しましょう。

```
・購入するユーザーのid
・購入する商品のid
```

を保存するテーブルです。
適宜各テーブルとのアソシエーションも追加もお願いいたします。

yutomentors reviewed (commented):
レビュー依頼ありがとうございます。
修正をお願いしております。ご確認いただけますと幸いです。
指摘が多く申し訳ないのですが、データベース設計はアプリ作成の土台となりますので丁寧に確認していきましょう。

修正後、コミットプッシュを確認いただき、再度**修正後のER図**を添付の上、**レビュー依頼フォーム**のご提出をいいたします。
お手数をお掛け致しますが、よろしくお願いいたします。

ご不明点や疑問点ございましたら、通話メンターにご相談ください！
どうぞ、よろしくお願いいたします。

hmeg2896 committed:
修正１

hmeg2896 commented:
https://gyazo.com/7aa969085c89943297f3e1ac0f4c9c64

chobi555 commented on README.md:
```diff
@@ -1,24 +1,97 @@
+
+
+
+## ordersテーブル
```
こちらのテーブルは誰がどの商品を購入したかを識別するテーブルですので、
usersテーブルとitemsテーブルの外部キーを定義しましょう。

chobi555 reviewed (commented):
レビュー依頼ありがとうございます。
コメント致しました。
確認後、修正をお願い致します。

hmeg2896 committed:
修正２

https://gyazo.com/ab65b0e5e3d2bf96eb67cdc5719070a7https://gyazo.com/ab65b0e5e3d2bf96eb67cdc5719070a7

mihot117 commented on README.md:
```diff
@@ -1,24 +1,96 @@
+| -------------------- | ---------- | ----------------------|
+| user_name            | string     | null: false           |
+| email                | string     | null: false           |
+|                      |            | unique: true          |
```
```ruby
| email | string | null: false |
|   |    | unique: true |
```

emailのuniqueを別行に記述してくださっていますが、
こちらはnullオプションと並列させていただいて問題ございませんので
以下のように記述をまとめていただきますようお願いいたします。

```ruby
| email | string | null: false, unique: true |
```

mihot117 commented on README.md:
```diff
@@ -1,24 +1,96 @@
+| postage_id             | integer       | null: false           |
+| prefecture_id          | integer       | null: false           |
+| region_id              | integer       | null: false           |
+| shipping_date_id       | integer       | null: false           |
```
active hashを用いて作成するカラムは合計で5つですが
こちらのテーブル内には6つ用意があるようにお見受けいたしました。
1つ不要なカラムがあるかと存じますので削除をお願いいたします。

mihot117 commented on README.md:
```diff
@@ -1,24 +1,96 @@
+| column                 | Type          | options               |
+| ---------------------- | ------------- | ----------------------|
+| postal_code            | string        | null: false           |
+| prefecture_id          | string        | null: false           |
```
```ruby
| prefecture_id | string | null: false |
```

上記の記述に関しましては、itemsテーブルの時と同様にinteger型で作成をお願いいたします。
id数値のみを管理するカラムであるためです。

mihot117 commented on README.md:
```diff
@@ -1,24 +1,96 @@
+| purchase_user_id   | string       | null: false           |
+| product_id         | string       | null: false           |
+| user               | reference    | foreign_key: true     |
+| item               | reference    | foreign_key: true     |
```
購入履歴テーブルでは、誰が・何を購入したのかが管理できれば問題ございませんので
user情報とitem情報のみを残し、以下は削除をお願いいたします。

```ruby
| purchase_user_id   | string       | null: false           |
| product_id         | string       | null: false           |
```

user情報、item情報ともに外部キーを取得してカラムに保存することができるため、新しく情報を入力する必要はございません。

mihot117 commented on README.md:
```diff
@@ -1,24 +1,96 @@
+has_many :items
+has_many :comments
+has_many :purchases
+has_many :orders
```
```ruby
has_many :purchases
```

purchasesテーブル内でuser情報を保存する必要はございませんので
上記は削除しておきましょう。

mihot117 commented on README.md:
```diff
@@ -1,24 +1,96 @@
+| user                   | references    | foreign_key: true     |  
 
-* Deployment instructions
+### アソシエーション
+belongs_to :user
+has_many :comments
+has_one :purchase
+has_many :orders
 
```
```ruby
has_one :purchase
```

purchasesテーブル内でitemsテーブルの外部キーは取得しませんので
上記は削除いただいて問題ございません。

```ruby
has_many :orders
```

また、ordersテーブルとのアソシエーションの見直しをお願いいたします。
現状では`has_many`で記述いただいていますが、has_manyですと
１つの商品に対して多数の購入履歴が存在していることになり、
一回売れた商品を何度も売ろうとしている状態になります。
商品は毎度1つしか存在しないかと存じますので、修正が必要です。

mihot117 commented on README.md:
```diff
@@ -1,24 +1,96 @@
+| user                   | references    | foreign_key: true     |  
+| order                  | references    | foreign_key: true     |
+
+### アソシエーション
+belongs_to :user
+belongs_to :order
+
```
```ruby
belongs_to :user
```

usersテーブルの外部キーは取得しませんので、上記は削除をお願いいたします。

mihot117 commented on README.md:
```diff
@@ -1,24 +1,96 @@
+| item               | reference    | foreign_key: true     |
+
+
+### アソシエーション
+belongs_to :user
+belongs_to :item
```
purchasesテーブルとのCDが存在していませんので追記をお願いいたします。

mihot117 reviewed (commented):
レビュー依頼ありがとうございます。

修正点確認させていただきました。
追加でコメントいたしましたので、ご確認いただけますと幸いです。
データベース設計は難易度が高いので修正が続いてしまいますが
ここでアプリの全体像を把握しておくと今後の実装がスムーズになりますので
もう一息、頑張ってまいりましょう。

修正後、再度レビュー依頼フォームのご提出ください。
また、修正後のER図のご提出も、プルリクエストに添付する形でお送りください。

ご対応のほど、よろしくお願いいたします。

hmeg2896 committed:
修正３

hmeg2896 commented:
https://gyazo.com/cf019fb59095221d31f1b4e40d4a91b9

mihot117 reviewed (commented):
レビュー依頼ありがとうございます！

修正点を確認いたしました。

**LGTM**とさせていただきます！
実装お疲れ様でした！

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に
次の実装のためのブランチを切って次の課題に取り組みましょう！

引き続き、開発頑張っていきましょう！応援しております。
